feat: add npm query cmd

ruyadorno opened the pull request:
Implemented Arborist `querySelectorAll` method that enables easier
retrieval of installed packages using a css-selector-like syntax.

It also includes a new command `npm query` that exposes that same logic
to users of the npm cli.

## References
Original RFC: https://github.com/npm/rfcs/pull/564

<!-- What / Why -->
<!-- Describe the request in detail. What it does and why it's being changed. -->


<!-- Examples:
  Related to #0
  Depends on #0
  Blocked by #0
  Fixes #0
  Closes #0
-->

ruyadorno committed:
feat: add npm query cmd

Implemented Arborist `querySelectorAll` method that enables easier
retrieval of installed packages using a css-selector-like syntax.

It also includes a new command `npm query` that exposes that same logic
to users of the npm cli.

Ref: https://github.com/npm/rfcs/pull/564

ruyadorno committed:
promisifies API and tree traversal

This will enable for implementing async pseudo-selectors such as
:outdate in the future.

ruyadorno committed:
fix: ci

- lint fix
- fix snapshot tests

ljharb commented on lib/utils/query-selector-all-response.js:
```diff
@@ -0,0 +1,30 @@
+      pkgid,
+    } = node.target
+
+    Object.assign(this, node.target.package)
```
```suggestion
  constructor ({ target }) {
    const {
      location,
      path,
      realpath,
      resolved,
      isLink,
      isWorkspace,
      pkgid,
    } = target

    Object.assign(this, target.package)
```

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,523 @@
+  #currentAstSelector = null
+
+  constructor (rootAstNode) {
+    this.#results = new Map()
```
```suggestion
  #results = new Map()
  #currentAstSelector = null

  constructor (rootAstNode) {
```

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,523 @@
+        acc.add(node)
+      }
+    }
+    return acc
```
```suggestion
    return new Set(rootAstNode.nodes.flatMap(n => this.#results.get(n)))
```
altho flatMap might not be available in all supported nodes :-(

wraithgar replied:
It is https://node.green/#ES2019-features-Array-prototype--flat--flatMap--Array-prototype-flatMap

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,523 @@
+      return (String(pkg[attribute] || '').match(/\w+/g) || []).includes(value)
+    },
+    '*=' ({ attribute, value, pkg }) {
+      return String(pkg[attribute] || '').indexOf(value) > -1
```
```suggestion
      return String(pkg[attribute] || '').includes(value)
```

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,523 @@
+    },
+  }))
+  const classesMap = new Map(Object.entries({
+    '.prod' (prevResults) {
+      return Promise.resolve(prevResults.filter(node =>
+        [...node.edgesIn].some(edge => edge.prod)))
+    },
```
```suggestion
    async '.prod' (prevResults) {
      return prevResults.filter(node =>
        [...node.edgesIn].some(edge => edge.prod))
    },
```
same on all the others, then it can't ever throw a sync exception

wraithgar replied:
Yep I'll audit for anything returning a promise to be an async for just this reason.

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,523 @@
+    }
+
+    return Promise.resolve(parentFound)
+  }
```
```suggestion
  const hasParent = async (node, compareNodes) => {
    if (parentCache.has(node) && parentCache.get(node).has(compareNodes)) {
      return true
    }
    const parentFound = compareNodes.some(compareNode => {
      // follows logical parent for link anscestors
      return (node.isTop && node.resolveParent) === compareNode ||
      // follows edges-in to check if they match a possible parent
      [...node.edgesIn].some(edge =>
        edge && edge.from === compareNode)
    })

    if (parentFound) {
      if (!parentCache.has(node)) {
        parentCache.set(node, new Set())
      }
      parentCache.get(node).add(compareNodes)
    }

    return parentFound
  }
```

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,523 @@
+          if (prop === arrayDelimiter) {
+            const newObjs = []
+            for (const obj of objs) {
+              if (Array.isArray(obj)) {
+                obj.forEach((i, index) => {
+                  newObjs.push(obj[index])
+                })
+              } else {
+                newObjs.push(obj)
+              }
```
```suggestion
              [].concat(obj).forEach((x) => {
                newObjs.push(x)
              })
```

wraithgar replied:
How is this not just `Array.flat()`?

ljharb replied:
ha, it might be :-p

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,523 @@
+    const classFn = classesMap.get(String(currentAstNode))
+    if (!classFn) {
+      throw Object.assign(
+        new Error(`\`${String(currentAstNode)}\` is not a supported class.`),
```
```suggestion
        new Error(`\`${currentAstNode}\` is not a supported class.`),
```

ljharb replied:
also, "class"

fritzy commented:
@darcyclarke I would love to know more about bugs you've mentioned encountering

wraithgar committed:
chore: refactor tests to use real npm

wraithgar committed:
fix: attribute selectors, workspace root, cleanup

Giant refactor:
  - adds case insensitivity to attribute selectors
  - fixes --include-workspace-root
  - fixes -w results
  - docs updates
  - consolidating state into the `results` object and passing that to
    the functions that the ast walker functions use.
  - optimizing and refactoring other loops
  - code consolidation and consistency between two different attribute
    selectors
  - Un-asyncify functions that don't do async operators.  We leave the
    exported fn async so we can add some in the future.
  - lots of other minor tweaks/cleanups

ljharb commented on lib/commands/query.js:
```diff
@@ -0,0 +1,93 @@
+const BaseCommand = require('../base-command.js')
+
+class QuerySelectorItem {
+  constructor (node) {
```
```suggestion
  constructor ({
    target: {
      pkgid: this.pkgid
      location: this.location,
      path: this.path,
      realpath: this.realpath,
      resolved: this.resolved,
      isLink: this.isLink,
      isWorkspace: this.isWorkspace,
    },
  }) {
```
i know this is a bit horrific but i think it does work :-p

wraithgar replied:
I think I'm gonna leave it as is.

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,512 @@
+            // otherwise just maps all currently found objs
+            // to the next prop from the lookup properties list,
+            // filters out any empty key lookup
+            objs = objs.map(obj => obj[prop]).filter(Boolean)
```
```suggestion
            objs = objs.flatMap(obj => obj[prop] || [])
```

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,512 @@
+              seenNodes.add(node)
+              return [...node.edgesIn]
+            })
+            .map(edge => edge.from)
+            .filter(Boolean)
```
```suggestion
            .flatMap(edge => [edge.from] || [])
```

wraithgar replied:
That won't work cause we'd just wrap the falsey edge.from in an array which is always truthy
```
> !![null]
true
```

wraithgar replied:
I think we just want `.flatMap(edge => edge.from || [])`

wraithgar replied:
This one likely needs deeper inspection cause it's already the result of a flatMap.

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,512 @@
+      return String(pkg[attribute] || '').includes(value)
+    },
+    '|=' ({ attribute, value, pkg }) {
+      return String(pkg[attribute] || '').split('-')[0] === value
```
```suggestion
      return String(pkg[attribute] || '').startsWith(`${value}-`)
```

wraithgar replied:
I assume you meant `${value}`

ljharb replied:
whoops yes, fixed thanks

ruyadorno commented on test/lib/commands/query.js:
```diff
@@ -0,0 +1,168 @@
+    },
+  })
+  await npm.exec('query', [':root, :root > *'])
+  t.matchSnapshot(joinedOutput(), 'should return root object')
```
the name of this assertion should probably be `should return root object and direct children` instead

wraithgar replied:
👍

ruyadorno commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -0,0 +1,543 @@
+const attributeOperators = {
+  // existence of the attribute
+  '' ({ attribute, value, pkg }) {
+    // TODO shouldn't this be some sort of hasProperty in case the value of the
```
I guess you're right! Something like `Object.hasOwn` is pb what we want here

wraithgar replied:
Yeah I'll get to this after I finish the case insensitivity stuff. Gotta come up w/ a test for it lol.

ruyadorno reviewed (commented):
awesome stuff @wraithgar!

I'm curious about the perf issues, I see you were using `p-map` before and I thought the improvements were due to that but it's no longer being used atm 🤔

wraithgar commented:
> awesome stuff @wraithgar!
> 
> I'm curious about the perf issues, I see you were using `p-map` before and I thought the improvements were due to that but it's no longer being used atm 🤔

Nothing under the hood was doing anything asynchronously.  Removing async functions altogether for functions that were completely synchronous made it faster.  We're also in the "make it right" phase.  Performance isn't really an issue right now, it takes less than a second to query the cli tree.  We can worry about performance next, after we "make it right".

wraithgar committed:
deps: @npmcli/query@1.1.0

wraithgar commented:
Refactor completed.  No more melting cpus on workspace / has queries.  include-workspace-root flag is obeyed.  Lots of other cleanups and code consolidations.  This may be ready to go.

wraithgar commented:
I spoke too soon. The depTypes/filterByType is not right and may never have been. Looking into that now.

wraithgar committed:
fix: proper depType functions

wraithgar commented:
`depType` filters fixed. Needs tests still, and one other tiny TODO having to do w/ number values in your packument.

ljharb commented on lib/commands/query.js:
```diff
@@ -17,6 +17,7 @@ class QuerySelectorItem {
     this.resolved = node.target.resolved
     this.isLink = node.target.isLink
     this.isWorkspace = node.target.isWorkspace
+    // XXX should we have dev, optional, devOptional?
```
dev and optional exist, so yes, devOptional would be great but i don't think it's a thing :-)

wraithgar replied:
It is an attribute on arborist nodes.

ljharb replied:
interesting. where does it come from, since package.json doesn't support devOptionalDependencies? or is it "optional deps of a dev dep"

wraithgar replied:
`devOptional is more internal to arborist, only adding `dev` `optional` `peer` from the node.

ljharb commented on workspaces/arborist/lib/query-selector-all.js:
```diff
@@ -392,24 +394,72 @@ const attributeMatch = (matcher, obj) => {
+    for (const edgeIn of node.edgesIn) {
+      if (edgeIn.type === type) {
+        found.push(node)
+        continue FBT_NODELOOP
```
lol i really hope there's another approach than GOTO here :-p

wraithgar replied:
Nope this is the most efficient way to break that loop. Learned this lesson when we super optimized the original npm audit server code.  Sometimes this is the best approach.

ljharb replied:
gross

wraithgar replied:
The better approach would be if `Set` objects had an equivalent to `Array.prototype.some`.  Outside of that we have to cast to an array which is slower and takes up more memory.

ljharb replied:
seems like at least it'd be sensible to make a `setSome` function abstraction to hide away this approach?

wraithgar replied:
That can be something we do in the future I suppose. I don't think "I don't like looking at this" warrants a refactor today.

wraithgar committed:
fix: implement deduped

wraithgar committed:
fix: look at everything with has operator

ruyadorno commented:
> Refactor completed. No more melting cpus on workspace / has queries. include-workspace-root flag is obeyed. Lots of other cleanups and code consolidations. This may be ready to go.

@wraithgar I'm curious on what was the issue with the melting cpus in the original implementation of the `has` queries 😅 How did it got fixed?

wraithgar commented:
@ruyadorno I'm not 100% certain to be honest. iirc there were a couple of places where the wrong values where being passed into `retrieveNodesFromParsedAst`. It was some kind of infinite loop that wasn't also infinite recursion. All I know for certain was that it was in rolling up all state into the `Results` object that the bug was fixed.

ruyadorno commented:
I see, sounds good! thanks for taking care and cleaning up all that 👍 I'm looking forward to this release 🤩

ruyadorno commented:
🥳

lorenzogrv commented:
Nice work guys. Thank you all :rocket:
